pumpkin/src/commands/cmd_say.rs

```diff
@@ -0,0 +1,111 @@
+
+    if all_args.is_empty() {
+        None
+    } else {
+        all_args.reverse();
+        Some(all_args.join(" "))
+    }
+}
+
+pub fn init_command_tree<'a>() -> CommandTree<'a> {
+    CommandTree::new(NAMES, DESCRIPTION).with_child(
+        argument(ARG_CONTENT, consume_arg_content).execute(&|sender, server, args| {
+            if let Some(content) = args.get("content") {
+                let content = parse_selectors(content, sender, server);
+                let message = &format!("[Console]: {content}");
+                let message = TextComponent::text(message).color_named(NamedColor::Blue);
+
+                server.broadcast_message(message.clone());
+                sender.send_message(message);
+            } else {
+                sender.send_message(
+                    TextComponent::text("Please provide a message: say [content]")
+                        .color_named(NamedColor::Red),
+                );
+            }
+
+            Ok(())
+        }),
+    )
+}
+
+fn parse_selectors(content: &str, sender: &CommandSender, server: &Server) -> String {
+    let mut final_message = String::new();
+
+    let tokens: Vec<&str> = content.split_whitespace().collect();
+
+    for token in tokens {
+        // TODO impl @e
+        let result = match token {
+            "@p" => {
+                let position = match sender {
+                    CommandSender::Player(p) => p.last_position.load(),
+                    _ => Vector3::new(0., 0., 0.),
+                };
+                vec![server
+                    .get_nearest_player(&position)
+                    .map_or_else(|| String::from("nobody"), |p| p.gameprofile.name.clone())]
+            }
+            "@r" => {
+                let online_players: Vec<String> = server
+                    .get_online_players()
+                    .map(|p| p.gameprofile.name.clone())
+                    .collect();
+
+                if online_players.is_empty() {
+                    vec![String::from("nobody")]
+                } else {
+                    vec![online_players[rand::random::<usize>() % online_players.len()].clone()]
+                }
+            }
+            "@s" => match sender {
+                CommandSender::Player(p) => vec![p.gameprofile.name.clone()],
+                _ => vec![String::from("console")],
+            },
+            "@a" => server
+                .get_online_players()
+                .map(|p| p.gameprofile.name.clone())
+                .collect::<Vec<_>>(),
+            "@here" => server
+                .get_online_players()
+                .map(|p| p.gameprofile.name.clone())
+                .collect::<Vec<_>>(),
+            _ => vec![token.to_string()],
+        };
+
+        // formatted player names
+        final_message.push_str(&format_player_names(&result));
+        final_message.push(' ');
+    }
+
+    final_message.trim_end().to_string()
+}
```

Comment:
We already have player arguments implemented, See for example gamemode command or kick command...
You can implement is easily by providing the target as an argument `.with_child(argument(ARG_TARGET, consume_arg_player)` and then parsing it ` let target = parse_arg_player(sender, server, ARG_TARGET, args)?;`

Reply:
The `say` command supports multiple target specifiers, it's not a command argument, but a way the content gets formatted